### DISCUSSION TITLE
How to build premake-core from source on MSYS2 UCRT64?

### DISCUSSION THREAD

**Discussion by ghost** [premake/premake-core#2348] [General]
I used the same command as in the PKGBUILD of MSYS2's premake package:

https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-premake/PKGBUILD

```
make -f Bootstrap.mak mingw CONFIG=release ARCH=x86_64
```

premake-bootstrap.exe is generated and it attempted to build premake5.exe, but I found it's building for `release_win32` and of course it will fail to link needed libraries, because my toolchain is x86_64 and not support multilib, so there is no x86 libraries.

The errors are like this:

```
C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: skipping incompatible C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/14.2.0/../../../libmingw32.a when searching for -lmingw32
```

I inspected the Bootstrap.mk file. I changed the command to be like the suggested in it:

```
make -f Bootstrap.mak mingw CONFIG=x64
```

Snippet:

```
none:
	@echo "Please do"
	@echo "   nmake -f Bootstrap.mak windows"
	@echo "or"
	@echo "   CC=mingw32-gcc mingw32-make -f Bootstrap.mak mingw CONFIG=x64"
	@echo "or"
	@echo "   make -f Bootstrap.mak HOST_PLATFORM"
	@echo "where HOST_PLATFORM is one of these:"
	@echo "   osx linux bsd"
	@echo ""
	@echo "To clean the source tree, run the same command by adding a '-clean' suffix to the target name."
		@echo "Example"
	@echo "   make -f Bootstrap.mak HOST_PLATFORM-clean"
```

But I found it's still generating for x86:

```
./build/bootstrap/premake_bootstrap --arch=x86 --os=windows --to=build/bootstrap --cc=mingw  gmake2
```

(extracted from the terminal output)

And `make` failed very quickly because `Makefile:78: *** "invalid configuration x64_win32".  Stop.`.

```
make -C build/bootstrap -j`getconf _NPROCESSORS_ONLN` config=x64_win32
```

What's going on? Even the embedded instruction in Bootstrap.mk is invalid.

Update: I found that you are hardcoding it to target x86:

```
mingw: mingw-clean
	mkdir -p build/bootstrap
	$(CC) -o build/bootstrap/premake_bootstrap -DPREMAKE_NO_BUILTIN_SCRIPTS -I"$(LUA_DIR)" -I"$(LUASHIM_DIR)" $(SRC) -lole32 -lversion
	./build/bootstrap/premake_bootstrap embed
	./build/bootstrap/premake_bootstrap --arch=$(PLATFORM) --os=windows --to=build/bootstrap --cc=mingw $(PREMAKE_OPTS) gmake2
	$(MAKE) -C build/bootstrap -j`getconf _NPROCESSORS_ONLN` config=$(CONFIG)_$(PLATFORM:x86=win32)
```

Why? And how can I change this behavior without having to edit Bootstrap.mk? How do you make binary builds for Windows? I found that you are also using MSYS2.

**Comment by ghost** [comment]
The correct command is:

```
make -f Bootstrap.mak mingw CONFIG=release PLATFORM=x64
```

The script set `--arch` to `$(PLATFORM)`, the environment variable `ARCH` is completely not being used. I have no idea why the PKGBUILD of MSYS2 still works and possible to produce binary packages.

Note: `x86_64` is invalid, only `x64` works for `PLATFORM`.

**Comment by samsinsane** [comment]
There's a lot to respond to here.

> I used the same command as in the PKGBUILD of MSYS2's premake package:

You haven't used the same command. It uses `PLATFORM="${_arch}"` where `$_arch` is set to `x64` when `$CARCH` is `x86_64`, you have `ARCH=x86_64` which is completely different.

> I inspected the Bootstrap.mk file. I changed the command to be like the suggested in it:

Looks like the `Bootstrap.mak` comments are out of date with the actual commands. It should be `PLATFORM=x64`.

> Update: I found that you are hardcoding it to target x86:

I don't see any hardcoding to `x86`? Unless you're referring to this `$(PLATFORM:x86=win32)`, this is just replacing `x86` with `win32`.

> How do you make binary builds for Windows?

With Visual Studio.

> I found that you are also using MSYS2.

I'm a bit confused by this comment, the only place we really use it is in the CI which shows you exactly how to build it:
https://github.com/premake/premake-core/blob/5c440ef2ca7ed4819b9a0d1e63e43aef575b518f/.github/workflows/ci-workflow.yml#L82-L120

---

> The correct command is:

Glad you found the correct command in the end.

> I have no idea why the PKGBUILD of MSYS2 still works and possible to produce binary packages.

As above, what you have and what PKGBUILD uses weren't the same.
